File: Data Structures & Algorithms/longest-increasing-subsequence/submission-1.cpp
class Solution {
public:
    int lengthOfLIS(vector<int>& nums) {
        if(nums.size()==1)return 1;

        int len=nums.size();
        vector<int>ans(nums.size(),1);

        for(int i=len-2;i>=0;i--) {
            for(int j=i+1;j<len;j++) {
                if(nums[j]>nums[i] && ans[j]>=ans[i])  {
                    ans[i]=ans[j]+1;
                }
            }
        }

        int lis=ans[0];
        for(int i=1;i<ans.size();i++) {
            if(ans[i]>lis)lis=ans[i];
        }
        return lis;
    }
};

File: Data Structures & Algorithms/products-of-array-discluding-self/submission-1.py
class Solution:
    def productExceptSelf(self, nums: List[int]) -> List[int]:
        prefixes=[]
        suffixes=()
        res=[0]*len(nums)

        prefix=1
        prefixes.append(1)
        for i in range(1,len(nums)):
            prefix=prefix*nums[i-1]
            prefixes.append(prefix)

        suffix=1
        suffixes=[1]*len(nums)
        for i in range(len(nums)-2,-1,-1):
            suffix=suffix*nums[i+1]
            suffixes[i]=suffix
        
        for i in range(len(nums)):
            res[i]=prefixes[i]*suffixes[i]
        
        return res
